### ISSUE TITLE
no-underscore-dangle rule is either broken or the documentation is wrong

### ISSUE THREAD

**Issue by krainboltgreene** [eslint/eslint#7064]
**What version of ESLint are you using?**
3.4.0

**What parser (default, Babel-ESLint, etc.) are you using?**
babel-eslint

**Please show your full configuration:**
https://gist.github.com/krainboltgreene/a8e6ae9c76869a75b087317c4b5882c2

**What did you do? Please include the actual source code causing the issue.**
Wrote:

``` javascript
const __foo = 1
```

**What did you expect to happen?**

No linting errors.

**What actually happened? Please include the actual, raw output from ESLint.**

<img width="695" alt="screen shot 2016-09-05 at 9 55 33 pm" src="https://cloud.githubusercontent.com/assets/334809/18262128/7d432f6a-73b3-11e6-997e-039e92860215.png">

Tried it in the demo tool, works fine. Nothing in the actual source of the rule suggests it would work with the documentation provided (IE making sure it's two underscores).

**Comment by ljharb** [comment]
The [documentation](http://eslint.org/docs/rules/no-underscore-dangle) implies to me that although you can *reference* something with double underscores (the RHS), you can't actually define a variable that way (the LHS) with the rule enabled. Note `var __proto__ = {};` marked as an error.

I'm not sure what the intention of the rule is, nor if the documentation should be improved, but imo this code absolutely should be an error with this rule enabled.

**Comment by krainboltgreene** [comment]
@ljharb So why is the demo marking it as lint free?

**Comment by ljharb** [comment]
`__filename` is magically provided to node modules by node itself; I'm assuming that's what the example was referring to.

I'm not sure why the demo allows it - that sounds like a bug in the demo.

**Comment by krainboltgreene** [comment]
Ah, I didn't know it was a special reference.

So this rule enforces:

No `_` (of any amount) before or after a variable or property name.
  - Unless it's underscore, a special library apparently.
  - Unless it's `__proto__` which is provided by the system.
  - Unless it's `__filename` which is provided by node.
  - Unless it's `__dirname` which is provided by node.

This feels like a very weird rule. Is there any flexibility? Private property syntax is a valuable notation and ES isn't going to adopt it's own anytime soon.

**Comment by ljharb** [comment]
Underscores aren't "private property syntax" - JS doesn't have privacy, although there is in fact a [state 1 proposal](https://github.com/tc39/proposal-private-fields) for actual privacy.

Barring that, everything reachable is public, whether there's an underscore in front of it or not - many [style guides](https://github.com/airbnb/javascript#naming--leading-underscore) discourage it because it creates a false sense of privacy that can lead to bugs. Of course, if you don't wish to comply with any rule, you're always free to disable it :-)

**Comment by krainboltgreene** [comment]
Yes @ljharb, I know all of that.

**Comment by kaicataldo** [comment]
Thanks for the issue report, @krainboltgreene.

Apologies if I'm missing something, but I'm not sure I'm following. It looks like `const __foo = 1` is being flagged as an issue, but you also mentioned:

> And the linter is OK with const __foo = 1

Those seem contradictory to me - mind clarifying?

As for the rule itself - I'm not too familiar with this rule, but looking at the source, it looks like it allows for two special cases: `__proto__` and `_`.

The rule only seems to check the left hand side of declarations, which is why `var file = __filename;` is allowed. My assumption is that the motivation for this is to prevent new declarations with a dangling underscore in the identifier (something you have control over), but allows usage of variables with a dangling underscore (something you may not have control over). So it seems to me like the rule is behaving as intended. I agree, though, that it's confusing.

The most confusing to me is that properties of `MemberExpressions` are checked on both sides, while other identifiers are only evaluated at the point of declaration. Some examples:
```js
/* "no-underscore-dangle": "error" */

var _foo = _baz(); //  1:5  error  Unexpected dangling '_' in '_foo' 
_foo = _baz() // No error
```
```js
/* "no-underscore-dangle": "error" */

foo = bar._baz(); // 1:7  error  Unexpected dangling '_' in '_baz'
foo = _bar.baz(); // No error
```

Maybe someone else on the team can clarify why the rule was written as it is.

**Comment by kaicataldo** [comment]
Ah, I think I see why - it's because there's no guarantee you have control over properties of objects, as you may not have any control over the naming scheme of those properties (i.e. they may be imported or the result of parsing JSON). We could definitely do a better job at making this clearer in our documentation.

**Comment by vitorbal** [comment]
ping @krainboltgreene, could you clarify what @kaicataldo mentioned above?

**Comment by krainboltgreene** [comment]
Sorry, the difference is between my machine vs the online tryeslint.

**Comment by vitorbal** [comment]
Just tried `const __foo = 1;` in the demo page and the `no-underscore-dangle` rule is triggered.

@kaicataldo looking at your second example, I would be inclined to say this is a bug, for the same reason you described. I went digging in the history of the rule definition file and I can't seem to find any explicit reason why we would have done it on purpose. It looks like we just forgot to account for  MemberExpressions that are part of a CallExpression.
What do you think, shall we open a bug ticket for that?

**Comment by kaicataldo** [comment]
@eslint/eslint-team Can we get some more eyes on this? If this is a bug, I can make a new issue to fix it!

**Comment by ilyavolodin** [comment]
I agree that what @kaicataldo mention seems strange to me. However, those issues are not related to the original issue filed by @krainboltgreene 
Our demo has been through a full re-write recently. And I just verified that `const __foo = 1` correctly displays a no-underscore-dangle error. So I believe that original issue is now resolved. @kaicataldo I think you should open a new issue for the cases that you mentioned.

**Comment by alberto** [comment]
Closing this issue as it looks like the question has been answered. Please feel free to visit us in the ESLint Gitter if you have any other issues!
